--- Day 12: Hot Springs ---
You finally reach the hot springs! You can see steam rising from secluded areas attached to the primary, ornate building.

As you turn to enter, the researcher stops you. "Wait - I thought you were looking for the hot springs, weren't you?" You indicate that this definitely looks like hot springs to you.

"Oh, sorry, common mistake! This is actually the onsen! The hot springs are next door."

You look in the direction the researcher is pointing and suddenly notice the massive metal helixes towering overhead. "This way!"

It only takes you a few more steps to reach the main gate of the massive fenced-off area containing the springs. You go through the gate and into a small administrative building.

"Hello! What brings you to the hot springs today? Sorry they're not very hot right now; we're having a lava shortage at the moment." You ask about the missing machine parts for Desert Island.

"Oh, all of Gear Island is currently offline! Nothing is being manufactured at the moment, not until we get more lava to heat our forges. And our springs. The springs aren't very springy unless they're hot!"

"Say, could you go up and see why the lava stopped flowing? The springs are too cold for normal operation, but we should be able to find one springy enough to launch you up there!"

There's just one problem - many of the springs have fallen into disrepair, so they're not actually sure which springs would even be safe to use! Worse yet, their condition records of which springs are damaged (your puzzle input) are also damaged! You'll need to help them repair the damaged records.

In the giant field just outside, the springs are arranged into rows. For each row, the condition records show every spring and whether it is operational (.) or damaged (#). This is the part of the condition records that is itself damaged; for some springs, it is simply unknown (?) whether the spring is operational or damaged.

However, the engineer that produced the condition records also duplicated some of this information in a different format! After the list of springs for a given row, the size of each contiguous group of damaged springs is listed in the order those groups appear in the row. This list always accounts for every damaged spring, and each number is the entire size of its contiguous group (that is, groups are always separated by at least one operational spring: #### would always be 4, never 2,2).

So, condition records with no unknown spring conditions might look like this:

#.#.### 1,1,3
.#...#....###. 1,1,3
.#.###.#.###### 1,3,1,6
####.#...#... 4,1,1
#....######..#####. 1,6,5
.###.##....# 3,2,1
However, the condition records are partially damaged; some of the springs' conditions are actually unknown (?). For example:

???.### 1,1,3
.??..??...?##. 1,1,3
?#?#?#?#?#?#?#? 1,3,1,6
????.#...#... 4,1,1
????.######..#####. 1,6,5
?###???????? 3,2,1
Equipped with this information, it is your job to figure out how many different arrangements of operational and broken springs fit the given criteria in each row.

In the first line (???.### 1,1,3), there is exactly one way separate groups of one, one, and three broken springs (in that order) can appear in that row: the first three unknown springs must be broken, then operational, then broken (#.#), making the whole row #.#.###.

The second line is more interesting: .??..??...?##. 1,1,3 could be a total of four different arrangements. The last ? must always be broken (to satisfy the final contiguous group of three broken springs), and each ?? must hide exactly one of the two broken springs. (Neither ?? could be both broken springs or they would form a single contiguous group of two; if that were true, the numbers afterward would have been 2,3 instead.) Since each ?? can either be #. or .#, there are four possible arrangements of springs.

The last line is actually consistent with ten different arrangements! Because the first number is 3, the first and second ? must both be . (if either were #, the first number would have to be 4 or higher). However, the remaining run of unknown spring conditions have many different ways they could hold groups of two and one broken springs:

?###???????? 3,2,1
.###.##.#...
.###.##..#..
.###.##...#.
.###.##....#
.###..##.#..
.###..##..#.
.###..##...#
.###...##.#.
.###...##..#
.###....##.#
In this example, the number of possible arrangements for each row is:

???.### 1,1,3 - 1 arrangement
.??..??...?##. 1,1,3 - 4 arrangements
?#?#?#?#?#?#?#? 1,3,1,6 - 1 arrangement
????.#...#... 4,1,1 - 1 arrangement
????.######..#####. 1,6,5 - 4 arrangements
?###???????? 3,2,1 - 10 arrangements
Adding all of the possible arrangement counts together produces a total of 21 arrangements.

For each row, count all of the different arrangements of operational and broken springs that meet the given criteria. What is the sum of those counts?

In [1]:
# let's start with building a verifier:
import re

def is_consistent(record, grouping):
    damaged = re.findall(r'#+', record)
    if len(damaged) == len(grouping):
        if all(len(r) == g for r, g in zip(damaged, grouping)):
            return True
    return False
is_consistent('#....######.#####.', (1,6,5))

True

In [2]:
# next a permutator:
from itertools import combinations

def num_solutions(dam_rec, grouping):
    tot_hashes = sum(grouping)
    num_q = len([1 for c in dam_rec if c == '?'])
    d_hashes = tot_hashes - len([1 for c in dam_rec if c == '#'])
    
    combos = combinations(range(num_q), d_hashes)
    combos = [''.join('#' if i in co  else '.' for i in range(num_q)) for co in combos]
    num_s = 0

    for co in combos:
        iter_co = iter(co)
        cand = re.sub(r'\?', lambda x: next(iter_co), dam_rec)
        num_s += is_consistent(cand, grouping)
    return num_s
num_solutions('?????##???????.?#??#', (2,2,4,1,1,2))

3

In [3]:
with open('input12.txt') as file:
    records = [line.split()[0] for line in file]
    file.seek(0)
    groupings = [tuple(int(num_str) for num_str in line.split()[1].split(',')) for line in file]
records[0], groupings[0]

('?????#?#?.?#?#.', (7, 3))

In [4]:
total_sum = 0
for rec, gr in zip(records, groupings):
    total_sum += num_solutions(rec, gr)
    print(total_sum, sep='')
print('finish!')
total_sum

2
4
13
19
83
85
88
92
100
101
105
106
116
134
136
142
146
165
168
174
186
189
191
193
241
243
245
279
282
283
363
365
369
372
378
380
383
385
388
400
404
407
415
421
423
427
433
438
443
445
448
452
483
489
495
497
498
504
515
517
559
565
567
568
580
582
586
588
657
663
667
671
673
679
680
686
688
694
704
734
737
749
755
757
759
760
766
784
798
803
824
830
831
835
841
848
852
865
871
872
874
877
879
882
891
896
901
904
906
908
912
921
925
933
935
936
948
951
961
964
986
994
1001
1025
1027
1029
1041
1051
1055
1061
1065
1067
1085
1088
1090
1094
1097
1099
1104
1114
1116
1117
1119
1129
1146
1148
1150
1156
1172
1179
1181
1187
1191
1203
1218
1222
1224
1225
1226
1227
1234
1242
1244
1247
1264
1265
1270
1280
1292
1295
1300
1302
1306
1309
1314
1321
1327
1335
1353
1364
1366
1368
1372
1376
1382
1394
1398
1399
1402
1403
1411
1458
1465
1470
1471
1472
1474
1476
1479
1482
1486
1487
1491
1492
1493
1494
1498
1502
1520
1561
1565
1572
1573
1575
1581
1599
1614
1615
1620
1621
1633
1636
1655
1657
1672
1706
17

7191

Your puzzle answer was 7191.

The first half of this puzzle is complete! It provides one gold star: *

--- Part Two ---
As you look out at the field of springs, you feel like there are way more springs than the condition records list. When you examine the records, you discover that they were actually folded up this whole time!

To unfold the records, on each row, replace the list of spring conditions with five copies of itself (separated by ?) and replace the list of contiguous groups of damaged springs with five copies of itself (separated by ,).

So, this row:

.# 1
Would become:

.#?.#?.#?.#?.# 1,1,1,1,1
The first line of the above example would become:

???.###????.###????.###????.###????.### 1,1,3,1,1,3,1,1,3,1,1,3,1,1,3
In the above example, after unfolding, the number of possible arrangements for some rows is now much larger:

???.### 1,1,3 - 1 arrangement
.??..??...?##. 1,1,3 - 16384 arrangements
?#?#?#?#?#?#?#? 1,3,1,6 - 1 arrangement
????.#...#... 4,1,1 - 16 arrangements
????.######..#####. 1,6,5 - 2500 arrangements
?###???????? 3,2,1 - 506250 arrangements
After unfolding, adding all of the possible arrangement counts together produces 525152.

Unfold your condition records; what is the new sum of possible arrangement counts?



In [5]:
# I'm gonna go out on a limb here and say that the brute force approach is not going to cut it anymore for part 2. 

# Here is a different approach:
# We take each consecutive string of question marks, and, given their length and whether they have # or . as neighbors, we can enumerate our possibilities: extend heaps from the left or right, combine heaps from the left and right, and build new heaps (one or multiple with various lengths). The number of options depends on the length of the sequence and neighbors present. And: if a sequence requires multiple dots in a row, there are multiple ways to get the same result. We can precalculate this and multiply by the number as options. 
# Then we can iterate through all the options. I had considered passing through the string first and only doing changes that don't multiply paths (other than the running count of ways to shift dots which does not increase compute down the line) before forking but omitted it for simplicity. The logic gets quite complex already.


In [6]:
rec_5 = ['?'.join([r] * 5) for r in records]
gr_5 = [g * 5 for g in groupings]
rec_5[0], gr_5[0]

('?????#?#?.?#?#.??????#?#?.?#?#.??????#?#?.?#?#.??????#?#?.?#?#.??????#?#?.?#?#.',
 (7, 3, 7, 3, 7, 3, 7, 3, 7, 3))

In [7]:
longest_q = len(max(re.findall(r'\?+', ''.join(rec_5)), key=len))
longest_q

85

In [8]:
from functools import cache
from math import comb
from collections import deque

def parse_str(rec):
    blocks = re.finditer(r'\?+|#+', rec)
    gr = []
    for b in blocks:
        if b[0].startswith('#'):
            gr.append(len(b[0]))
        elif b[0].startswith('?'):
            left_ext = right_ext = False
            if b.start() > 0 and rec[b.start() - 1] == '#':
                left_ext = True
            if b.end() < len(rec) and rec[b.end()] == '#':
                right_ext = True
            gr.append((left_ext, right_ext, len(b[0])))
    return tuple(gr)
parse_str('.??..??...?##.')

((False, False, 2), (False, False, 2), (False, True, 1), 2)

In [9]:
# Let's build up our set of possibilities like this:
# If hashes on left and right, one possibility is combination with value L + R + q_len
# If hashes on either side, one possibility is filling up with up to q_len hashes for value of L/R + q_len
# We can also fill it up with smaller values, leaving x leftover-dots
# For the purposes of building new heaps, we are going to assume we have a sequence of question marks boundaried by white dots (or start/end) on both sides (if not we subtract a boundary dot as shown above). Now each additional heap will require one additional boundary dot, so eg if we have . boundaried ??? with len x, we can have one heap len x, two heaps with combined len x-1, three with combined len x-2 etc up to a maximum of len(q) >= 2n - 1 size 1 heaps
# Now leftover dots are given by the reverse: if within a .-boundaried q-space we have n heaps of combined size m, the number of leftover dots is o = len(q) - m - (n-1). The number of places they can attach to is 2 (on the left and on the right) + n-1 (the number of inside boundaries). 
# For the number of permutations where the leftover dots can be placed then we have stars and bars - drawing o dots from n + 1 possible placements with replacement, giving comb(o + n, o)

@cache
def num_comb(q, seq, right_connect=False):
    l_ext, r_ext, q_len = q
    assert not l_ext
    if right_connect: # with right-connect, the last seq_value hugs the right hand side
        q_len -= (seq[-1] + 1)
        seq.pop()
    elif r_ext:
        q_len -= 1

    free_dots = q_len - sum(seq) - len(seq) + 1
    return comb(free_dots + len(seq), free_dots)

In [23]:
# Let's look at the heap making/sizing logic in more detail. Roughly go from left to right. We have to do validation: If rec is incompatible with sol, we return 0. We don't have to do the full calc, we can return sub-strings to be popped into the q
# In detail, we pop an element from the record. If it is a #, we check if we need to grow it. If yes, we try to grow it from the next ? segment and hand it back to the q, if we cannot we return 0. If we don't need to (read: must not) grow it, we procceed to the next segment
# By contrast, if our element is a ? segment, we first look at extending a heap to the left. We don't actually need to extend, since that logic was covered previously. But we do need to check if we need to add a . on the left, that is close the heap to the left (remove left_ext, and decrease value by 1).
# Next, we check if it is right-extendable. If yes, we try out all reasonable values to extend to the right. 
# Next, we fill out the ???. On one end, we take the null option (simply delete the segment) and pass that back to the q (). If the segment is right extendable, we instead extend the right segment to the needed length. On the other hand, we try to make a new heap with the value from the solution and pass that back. Let's try to rely on caching here and not worry too much about doing efficient substitutions before exploding the tree. Im not sure if caching will be possible or we have too many combinations but let's see

@cache
def build_paths(rec, sol, pre_count=1):
    if not rec or not sol:
        if sol:
            return None
        if sum([x for x in rec if isinstance(x, int)]):
            return None
        return pre_count
    r_q = deque(rec)
    s_q = deque(sol)

    r = r_q.popleft()
    if isinstance(r, int): # popped in heap
        s = s_q.popleft()
        if (d:=s-r) < 0:
            return None
        elif d == 0:
            return [(tuple(r_q), tuple(s_q), pre_count)]
        else: # see if we can fill it up from the right
            l_heap = r
            if not r_q:
                return None
            r = r_q.popleft()
            if isinstance(r, tuple) and r[0]:
                if r[2] >= d + 2: # we can extend the heap and have leftover-qs
                    r_q.appendleft((False, r[1], r[2] - d - 1))
                    return [(tuple(r_q), tuple(s_q), pre_count)]
                elif r[2] == d + 1: # we can extend the heap without leftover-qs
                    return [(tuple(r_q), tuple(s_q), pre_count)]
                elif r[2] == d: # only soluble if r doesn't wrap to the right
                    if not r[1]:
                        return [(tuple(r_q), tuple(s_q), pre_count)]
                elif r[2] < d: # only potentially soluble if r does wrap to the right (fill and combine)
                    if r[1]:
                        next_heap = r_q.popleft()
                        l_heap += r[2] + next_heap
                        r_q.appendleft(l_heap)
                        s_q.appendleft(s)
                        return [(tuple(r_q), tuple(s_q), pre_count)]
            return None
    else: # start with q_seq
        if r[0]: # close left side since any left side mismatches should've been corrected on previous passes
            if r[2] > 1:
                r_q.appendleft((False, r[1], r[2] - 1))
            return [(tuple(r_q),tuple(s_q), pre_count)]
        if not r[1]: # no right extend
            sum_heaps = 0
            s_left = deque()
            s_right = s_q.copy()
            n_div = -1
            return_values = [(tuple(r_q), tuple(s_q), pre_count)]
            while s_right:
                heap = s_right.popleft()
                s_left.append(heap)
                sum_heaps += heap
                n_div += 1
                space_need = n_div + sum_heaps
                if space_need > r[2]:
                    break
                return_values.append((tuple(r_q), tuple(s_right), num_comb(r, tuple(s_left)) * pre_count))
            return return_values
        else: # right extend possible - fill to the right with various values and close it up 
            return_values = []
            space_need = 0
            max_fill = 0
            for s in s_q:
                possible_fill = min(s - r_q[0], r[2] - space_need)
                max_fill = max(max_fill, possible_fill)
                space_need += 1 + s
                if space_need >= r[2]:
                    break
            for f in range(0, max_fill + 1):
                mod_r_q = r_q.copy()
                mod_r_q[0] = r_q[0] + f
                mod_q = (r[0], False, r[2] - f - 1)
                if mod_q[2] > 0:
                    mod_r_q.appendleft(mod_q)
                return_values.append((tuple(mod_r_q), tuple(s_q), pre_count))
            return return_values 
num_solutions(parse_str('.??..??...?##.'), (1,1,3))

0

In [11]:
def is_valid(return_value):
    try:
        for r_q, s_q, count in return_value:
            for v1, v2 in zip(r_q, r_q[1:]):
                if isinstance(v1, tuple) and isinstance(v2, tuple):
                    if v1[1]:
                        return False
    except TypeError:
        pass
    return True

In [22]:
def rec_count(string, grouping,debug=False):
    queue = [(parse_str(string), grouping)]
    sum_total = 0
    while queue:
        if debug:
            x = queue.pop()
            new = build_paths(*x)
            print(f'{x=}')
            print(f'{new=}')
            print('------')
            if not is_valid(new):
                print('INVALID:')
                print(x)
                print('------')
                print(new)
        else:
            new = build_paths(*queue.pop())
        if isinstance(new, int):
            sum_total += new
        elif isinstance(new, list):
            queue.extend(new)
    return sum_total
            
rec_count('?'.join(['??#??????#???.?'] * 5), (4,3)*5)

186624

In [25]:
# testing:
for rec, gr in zip(records, groupings):
    if num_solutions(rec, gr) != rec_count(rec, gr):
        print(f'Error for puzzle {rec} {gr}:')
        print(f'True output {num_solutions(rec, gr)} but got {rec_count(rec, gr)}')

In [28]:
total_sum = 0
for i, (rec, gr) in enumerate(zip(rec_5, gr_5)):
    total_sum += rec_count(rec, gr)
    print(f'{i}/{len(rec_5)}')
print('finish!')
total_sum

0/1000
1/1000
2/1000
3/1000
4/1000
5/1000
6/1000
7/1000
8/1000
9/1000
10/1000
11/1000
12/1000
13/1000
14/1000
15/1000
16/1000
17/1000
18/1000
19/1000
20/1000
21/1000
22/1000
23/1000
24/1000
25/1000
26/1000
27/1000
28/1000
29/1000
30/1000
31/1000
32/1000
33/1000
34/1000
35/1000
36/1000
37/1000
38/1000
39/1000
40/1000
41/1000
42/1000
43/1000
44/1000
45/1000
46/1000
47/1000
48/1000
49/1000
50/1000
51/1000
52/1000
53/1000
54/1000
55/1000
56/1000
57/1000
58/1000
59/1000
60/1000
61/1000
62/1000
63/1000
64/1000
65/1000
66/1000
67/1000
68/1000
69/1000
70/1000
71/1000
72/1000
73/1000
74/1000
75/1000
76/1000
77/1000
78/1000
79/1000
80/1000
81/1000
82/1000
83/1000
84/1000
85/1000
86/1000
87/1000
88/1000
89/1000
90/1000
91/1000
92/1000
93/1000
94/1000
95/1000
96/1000
97/1000
98/1000
99/1000
100/1000
101/1000
102/1000
103/1000
104/1000
105/1000
106/1000
107/1000
108/1000
109/1000
110/1000
111/1000
112/1000
113/1000
114/1000
115/1000
116/1000
117/1000
118/1000
119/1000
120/1000
121/1000
122/1000
123

6512849198636

In [34]:
# ones that took a long time (index): 68, 524, 583, total runtime 2m36s
records[583], groupings[583]

('??#?.??.??.??????#?', (3, 1, 2, 2, 2))

In [12]:
import psutil, os

process = psutil.Process(os.getpid())
print(process.memory_info().rss / 1024 ** 2, "MB")

193.5 MB


That's the right answer! You are one gold star closer to restoring snow operations.

You have completed Day 12! You can [Share] this victory or [Return to Your Advent Calendar].

In [38]:
# Well, that one was a struggle, but we limped over the finish line. Let's improve on the whole caching aspect (after a convo with Claude):
@cache
def build_paths(rec, sol):
    if not rec or not sol:
        if sol:
            return None
        if sum([x for x in rec if isinstance(x, int)]):
            return None
        return 1
    r_q = deque(rec)
    s_q = deque(sol)

    r = r_q.popleft()
    if isinstance(r, int): # popped in heap
        s = s_q.popleft()
        if (d:=s-r) < 0:
            return None
        elif d == 0:
            return [(tuple(r_q), tuple(s_q), 1)]
        else: # see if we can fill it up from the right
            l_heap = r
            if not r_q:
                return None
            r = r_q.popleft()
            if isinstance(r, tuple) and r[0]:
                if r[2] >= d + 2: # we can extend the heap and have leftover-qs
                    r_q.appendleft((False, r[1], r[2] - d - 1))
                    return [(tuple(r_q), tuple(s_q), 1)]
                elif r[2] == d + 1: # we can extend the heap without leftover-qs
                    return [(tuple(r_q), tuple(s_q), 1)]
                elif r[2] == d: # only soluble if r doesn't wrap to the right
                    if not r[1]:
                        return [(tuple(r_q), tuple(s_q), 1)]
                elif r[2] < d: # only potentially soluble if r does wrap to the right (fill and combine)
                    if r[1]:
                        next_heap = r_q.popleft()
                        l_heap += r[2] + next_heap
                        r_q.appendleft(l_heap)
                        s_q.appendleft(s)
                        return [(tuple(r_q), tuple(s_q), 1)]
            return None
    else: # start with q_seq
        if r[0]: # close left side since any left side mismatches should've been corrected on previous passes
            if r[2] > 1:
                r_q.appendleft((False, r[1], r[2] - 1))
            return [(tuple(r_q),tuple(s_q), 1)]
        if not r[1]: # no right extend
            sum_heaps = 0
            s_left = deque()
            s_right = s_q.copy()
            n_div = -1
            return_values = [(tuple(r_q), tuple(s_q), 1)]
            while s_right:
                heap = s_right.popleft()
                s_left.append(heap)
                sum_heaps += heap
                n_div += 1
                space_need = n_div + sum_heaps
                if space_need > r[2]:
                    break
                return_values.append((tuple(r_q), tuple(s_right), num_comb(r, tuple(s_left))))
            return return_values
        else: # right extend possible - fill to the right with various values and close it up 
            return_values = []
            space_need = 0
            max_fill = 0
            for s in s_q:
                possible_fill = min(s - r_q[0], r[2] - space_need)
                max_fill = max(max_fill, possible_fill)
                space_need += 1 + s
                if space_need >= r[2]:
                    break
            for f in range(0, max_fill + 1):
                mod_r_q = r_q.copy()
                mod_r_q[0] = r_q[0] + f
                mod_q = (r[0], False, r[2] - f - 1)
                if mod_q[2] > 0:
                    mod_r_q.appendleft(mod_q)
                return_values.append((tuple(mod_r_q), tuple(s_q), 1))
            return return_values 

def rec_count(string, grouping,debug=False):
    queue = [(parse_str(string), grouping, 1)]
    sum_total = 0
    while queue:
        # if debug:
        #     # x = queue.pop()
        #     # new = build_paths(*x)
        #     # print(f'{x=}')
        #     # print(f'{new=}')
        #     # print('------')
        #     # if not is_valid(new):
        #     #     print('INVALID:')
        #     #     print(x)
        #     #     print('------')
        #     #     print(new)
        # else:
        n_rec, n_gr, pre_count = queue.pop()
        new = build_paths(n_rec, n_gr)
        if isinstance(new, int):
            sum_total += new * pre_count
        elif isinstance(new, list):
            new = [(rec, gr, num_paths * pre_count) for rec, gr, num_paths in new]
            queue.extend(new)
    return sum_total


In [ ]:
# testing:
for rec, gr in zip(records, groupings):
    if num_solutions(rec, gr) != rec_count(rec, gr):
        print(f'Error for puzzle {rec} {gr}:')
        print(f'True output {num_solutions(rec, gr)} but got {rec_count(rec, gr)}')

In [40]:
total_sum = 0
for i, (rec, gr) in enumerate(zip(rec_5, gr_5)):
    total_sum += rec_count(rec, gr)
    print(f'{i}/{len(rec_5)}')
print('finish!')
total_sum

0/1000
1/1000
2/1000
3/1000
4/1000
5/1000
6/1000
7/1000
8/1000
9/1000
10/1000
11/1000
12/1000
13/1000
14/1000
15/1000
16/1000
17/1000
18/1000
19/1000
20/1000
21/1000
22/1000
23/1000
24/1000
25/1000
26/1000
27/1000
28/1000
29/1000
30/1000
31/1000
32/1000
33/1000
34/1000
35/1000
36/1000
37/1000
38/1000
39/1000
40/1000
41/1000
42/1000
43/1000
44/1000
45/1000
46/1000
47/1000
48/1000
49/1000
50/1000
51/1000
52/1000
53/1000
54/1000
55/1000
56/1000
57/1000
58/1000
59/1000
60/1000
61/1000
62/1000
63/1000
64/1000
65/1000
66/1000
67/1000
68/1000
69/1000
70/1000
71/1000
72/1000
73/1000
74/1000
75/1000
76/1000
77/1000
78/1000
79/1000
80/1000
81/1000
82/1000
83/1000
84/1000
85/1000
86/1000
87/1000
88/1000
89/1000
90/1000
91/1000
92/1000
93/1000
94/1000
95/1000
96/1000
97/1000
98/1000
99/1000
100/1000
101/1000
102/1000
103/1000
104/1000
105/1000
106/1000
107/1000
108/1000
109/1000
110/1000
111/1000
112/1000
113/1000
114/1000
115/1000
116/1000
117/1000
118/1000
119/1000
120/1000
121/1000
122/1000
123

6512849198636

In [ ]:
# well that actually became slower.. I think the caching just doesnt save much time if it just simplifies a single step. Let's change that:

@cache
def build_paths(rec, sol):
    if not rec or not sol:
        if sol:
            return 0
        if sum([x for x in rec if isinstance(x, int)]):
            return 0
        return 1
    r_q = deque(rec)
    s_q = deque(sol)

    r = r_q.popleft()
    if isinstance(r, int): # popped in heap
        s = s_q.popleft()
        if (d:=s-r) < 0:
            return 0
        elif d == 0:
            return build_paths(tuple(r_q), tuple(s_q))
        else: # see if we can fill it up from the right
            l_heap = r
            if not r_q:
                return 0
            r = r_q.popleft()
            if isinstance(r, tuple) and r[0]:
                if r[2] >= d + 2: # we can extend the heap and have leftover-qs
                    r_q.appendleft((False, r[1], r[2] - d - 1))
                    return build_paths(tuple(r_q), tuple(s_q))
                elif r[2] == d + 1: # we can extend the heap without leftover-qs
                    return build_paths(tuple(r_q), tuple(s_q))
                elif r[2] == d: # only soluble if r doesn't wrap to the right
                    if not r[1]:
                        return build_paths(tuple(r_q), tuple(s_q))
                elif r[2] < d: # only potentially soluble if r does wrap to the right (fill and combine)
                    if r[1]:
                        next_heap = r_q.popleft()
                        l_heap += r[2] + next_heap
                        r_q.appendleft(l_heap)
                        s_q.appendleft(s)
                        return build_paths(tuple(r_q), tuple(s_q))
            return 0
    else: # start with q_seq
        if r[0]: # close left side since any left side mismatches should've been corrected on previous passes
            if r[2] > 1:
                r_q.appendleft((False, r[1], r[2] - 1))
            return build_paths(tuple(r_q), tuple(s_q))
        if not r[1]: # no right extend
            sum_heaps = 0
            s_left = deque()
            s_right = s_q.copy()
            n_div = -1
            paths = [(tuple(r_q), tuple(s_q), 1)]
            while s_right:
                heap = s_right.popleft()
                s_left.append(heap)
                sum_heaps += heap
                n_div += 1
                space_need = n_div + sum_heaps
                if space_need > r[2]:
                    break
                paths.append((tuple(r_q), tuple(s_right), num_comb(r, tuple(s_left))))
            return sum((build_paths(r, s) * c for r, s, c in paths))
        else: # right extend possible - fill to the right with various values and close it up 
            paths = []
            space_need = 0
            max_fill = 0
            for s in s_q:
                possible_fill = min(s - r_q[0], r[2] - space_need)
                max_fill = max(max_fill, possible_fill)
                space_need += 1 + s
                if space_need >= r[2]:
                    break
            for f in range(0, max_fill + 1):
                mod_r_q = r_q.copy()
                mod_r_q[0] = r_q[0] + f
                mod_q = (r[0], False, r[2] - f - 1)
                if mod_q[2] > 0:
                    mod_r_q.appendleft(mod_q)
                paths.append((tuple(mod_r_q), tuple(s_q)))
            return sum((build_paths(r, s) for r, s in paths))

In [44]:
# testing:
for rec, gr in zip(records, groupings):
    if num_solutions(rec, gr) != build_paths(parse_str(rec), gr):
        print(f'Error for puzzle {rec} {gr}:')
        print(f'True output {num_solutions(rec, gr)} but got {build_paths(parse_str(rec), gr)}')

In [11]:
total_sum = 0
for i, (rec, gr) in enumerate(zip(rec_5, gr_5)):
    total_sum += build_paths(parse_str(rec), gr)
print('finish!')
total_sum

finish!


6512849198636

In [ ]:
# Now that was fast.